### ISSUE TITLE
Why Initialize a Neural Network with Random Weights?

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#443]
Why Initialize a Neural Network with Random Weights?<br>
<br>
https://ift.tt/2voK8gl<br>
<br>
Jason Brownlee<br>
<br>
<p>The weights of artificial neural networks must be initialized to small random numbers.</p>
<p>This is because this is an expectation of the stochastic optimization algorithm used to train the model, called stochastic gradient descent.</p>
<p>To understand this approach to problem solving, you must first understand the role of nondeterministic and randomized algorithms as well as the need for stochastic optimization algorithms to harness randomness in their search process.</p>
<p>In this post, you will discover the full background as to why neural network weights must be randomly initialized.</p>
<p>After reading this post, you will know:</p>
<ul>
<li>About the need for nondeterministic and randomized algorithms for challenging problems.</li>
<li>The use of randomness during initialization and search in stochastic optimization algorithms.</li>
<li>That stochastic gradient descent is a stochastic optimization algorithm and requires the random initialization of network weights.</li>
</ul>
<p>Let&rsquo;s get started.</p>
<div class="wp-caption">
<img src="https://machinelearningmastery.com/wp-content/uploads/2018/08/Why-Initialize-a-Neural-Network-with-Random-Weights.jpg" alt="Why Initialize a Neural Network with Random Weights?" width="640" class="wp-image-5902" height="424">
<p class="wp-caption-text">Why Initialize a Neural Network with Random Weights?<br>
Photo by <a href="https://www.flickr.com/photos/37195641@N03/7086827121/">lwtt93</a>, some rights reserved.</p>
</div>
<h2>Overview</h2>
<p>This post is divided into 4 parts; they are:</p>
<ol>
<li>Deterministic and Non-Deterministic Algorithms</li>
<li>Stochastic Search Algorithms</li>
<li>Random Initialization in Neural Networks</li>
<li>Initialization Methods</li>
</ol>
<h2>Deterministic and Non-Deterministic Algorithms</h2>
<p>Classical algorithms are deterministic.</p>
<p>An example is an algorithm to sort a list.</p>
<p>Given an unsorted list, the sorting algorithm, say bubble sort or quick sort, will systematically sort the list until you have an ordered result. Deterministic means that each time the algorithm is given the same list, it will execute in exactly the same way. It will make the same moves at each step of the procedure.</p>
<p>Deterministic algorithms are great as they can make guarantees about best, worst, and average running time. The problem is, they are not suitable for all problems.</p>
<p>Some problems are hard for computers. Perhaps because of the number of combinations; perhaps because of the size of data. They are so hard because a deterministic algorithm cannot be used to solve them efficiently. The algorithm may run, but will continue running until the heat death of the universe.</p>
<p>An alternate solution is to use <a href="https://en.wikipedia.org/wiki/Nondeterministic_algorithm">nondeterministic algorithms</a>. These are algorithms that use elements of <a href="https://en.wikipedia.org/wiki/Randomized_algorithm">randomness</a> when making decisions during the execution of the algorithm. This means that a different order of steps will be followed when the same algorithm is rerun on the same data.</p>
<p>They can rapidly speed up the process of getting a solution, but the solution will be approximate, or &ldquo;<em>good</em>,&rdquo; but often not the &ldquo;<em>best</em>.&rdquo; Nondeterministic algorithms often cannot make strong guarantees about running time or the quality of the solution found.</p>
<p>This is often fine as the problems are so hard that any good solution will often be satisfactory.</p>
<h2>Stochastic Search Algorithms</h2>
<p>Search problems are often very challenging and require the use of nondeterministic algorithms that make heavy use of randomness.</p>
<p>The algorithms are not random per se; instead they make careful use of randomness. They are random within a bound and are referred to as <a href="https://en.wikipedia.org/wiki/Stochastic_optimization">stochastic algorithms</a>.</p>
<p>The incremental, or step-wise, nature of the search often means the process and the algorithms are referred to as an optimization from an initial state or position to a final state or position. For example, stochastic optimization problem or a stochastic optimization algorithm.</p>
<p>Some examples include the genetic algorithm, simulated annealing, and stochastic gradient descent.</p>
<p>The search process is incremental from a starting point in the space of possible solutions toward some good enough solution.</p>
<p>They share common features in their use of randomness, such as:</p>
<ul>
<li>Use of randomness during initialization.</li>
<li>Use of randomness during the progression of the search.</li>
</ul>
<p>We know nothing about the structure of the search space. Therefore, to remove bias from the search process, we start from a randomly chosen position.</p>
<p>As the search process unfolds, there is a risk that we are stuck in an unfavorable area of the search space. Using randomness during the search process gives some likelihood of getting unstuck and finding a better final candidate solution.</p>
<p>The idea of getting stuck and returning a less-good solution is referred to as getting stuck in a local optima.</p>
<p>These two elements of random initialization and randomness during the search work together.</p>
<p>They work together better if we consider any solution found by the search as provisional, or a candidate, and that the search process can be performed multiple times.</p>
<p>This gives the stochastic search process multiple opportunities to start and traverse the space of candidate solutions in search of a better candidate solution&ndash;a so-called global optima.</p>
<p>The navigation of the space of candidate solutions is often described using the analogy of a one- or two-landscape of mountains and valleys (e.g. like a <a href="https://en.wikipedia.org/wiki/Fitness_landscape">fitness landscape</a>). If we are maximizing a score during the search, we can think of small hills in the landscape as a local optima and the largest hills as the global optima.</p>
<p>This is a fascinating area of research, an area where I have some background. For example, see my book:</p>
<ul>
<li><a href="http://cleveralgorithms.com/nature-inspired/index.html">Clever Algorithms: Nature-Inspired Programming Recipes</a></li>
</ul>
<h2>Random Initialization in Neural Networks</h2>
<p>Artificial neural networks are trained using a stochastic optimization algorithm called stochastic gradient descent.</p>
<p>The algorithm uses randomness in order to find a good enough set of weights for the specific mapping function from inputs to outputs in your data that is being learned. It means that your specific network on your specific training data will fit a different network with a different model skill each time the training algorithm is run.</p>
<p>This is a feature, not a bug.</p>
<p>I write about this issue more in the post:</p>
<ul>
<li><a href="https://machinelearningmastery.com/randomness-in-machine-learning/">Embrace Randomness in Machine Learning</a></li>
</ul>
<p>As described in the previous section, stochastic optimization algorithms such as stochastic gradient descent use randomness in selecting a starting point for the search and in the progression of the search.</p>
<p>Specifically, stochastic gradient descent requires that the weights of the network are initialized to small random values (random, but close to zero, such as in [0.0, 0.1]). Randomness is also used during the search process in the shuffling of the training dataset prior to each epoch, which in turn results in differences in the gradient estimate for each batch.</p>
<p>You can learn more about stochastic gradient descent in this post:</p>
<ul>
<li><a href="https://machinelearningmastery.com/gentle-introduction-mini-batch-gradient-descent-configure-batch-size/">A Gentle Introduction to Mini-Batch Gradient Descent and How to Configure Batch Size</a></li>
</ul>
<p>The progression of the search or learning of a neural network is referred to as convergence. The discovering of a sub-optimal solution or local optima is referred to as premature convergence.</p>
<blockquote>
<p>Training algorithms for deep learning models are usually iterative in nature and thus require the user to specify some initial point from which to begin the iterations. Moreover, training deep models is a sufficiently difficult task that most algorithms are strongly affected by the choice of initialization.</p>
</blockquote>
<p>&mdash; Page 301, <a href="https://amzn.to/2H5wjfg">Deep Learning</a>, 2016.</p>
<p>The most effective way to evaluate the skill of a neural network configuration is to repeat the search process multiple times and report the average performance of the model over those repeats. This gives the configuration the best chance to search the space from multiple different sets of initial conditions. Sometimes this is called a multiple restart or multiple-restart search.</p>
<p>You can learn more about the effective evaluation of neural networks in this post:</p>
<ul>
<li><a href="https://machinelearningmastery.com/evaluate-skill-deep-learning-models/">How to Evaluate the Skill of Deep Learning Models</a></li>
</ul>
<h3>Why Not Set Weights to Zero?</h3>
<p>We can use the same set of weights each time we train the network; for example, you could use the values of 0.0 for all weights.</p>
<p>In this case, the equations of the learning algorithm would fail to make any changes to the network weights, and the model will be stuck. It is important to note that the bias weight in each neuron is set to zero by default, not a small random value.</p>
<p>Specifically, nodes that are side-by-side in a hidden layer connected to the same inputs must have different weights for the learning algorithm to update the weights.</p>
<p>This is often referred to as the need to break symmetry during training.</p>
<blockquote>
<p>Perhaps the only property known with complete certainty is that the initial parameters need to &ldquo;break symmetry&rdquo; between different units. If two hidden units with the same activation function are connected to the same inputs, then these units must have different initial parameters. If they have the same initial parameters, then a deterministic learning algorithm applied to a deterministic cost and model will constantly update both of these units in the same way.</p>
</blockquote>
<p>&mdash; Page 301, <a href="https://amzn.to/2H5wjfg">Deep Learning</a>, 2016.</p>
<h3>When to Initialize to the Same Weights?</h3>
<p>We could use the same set of random numbers each time the network is trained.</p>
<p>This would not be helpful when evaluating network configurations.</p>
<p>It may be helpful in order to train the same final set of network weights given a training dataset in the case where a model is being used in a production environment.</p>
<p>You can learn more about fixing the random seed for neural networks developed with Keras in this post:</p>
<ul>
<li><a href="https://machinelearningmastery.com/reproducible-results-neural-networks-keras/">How to Get Reproducible Results with Keras</a></li>
</ul>
<h2>Initialization Methods</h2>
<p>Traditionally, the weights of a neural network were set to small random numbers.</p>
<p>The initialization of the weights of neural networks is a whole field of study as the careful initialization of the network can speed up the learning process.</p>
<p>Modern deep learning libraries, such as Keras, offer a host of network initialization methods, all are variations of initializing the weights with small random numbers.</p>
<p>For example, the current methods are available in Keras at the time of writing for all network types:</p>
<ul>
<li>
<strong>Zeros</strong>: Initializer that generates tensors initialized to 0.</li>
<li>
<strong>Ones</strong>: Initializer that generates tensors initialized to 1.</li>
<li>
<strong>Constant</strong>: Initializer that generates tensors initialized to a constant value.</li>
<li>
<strong>RandomNormal</strong>: Initializer that generates tensors with a normal distribution.</li>
<li>
<strong>RandomUniform</strong>: Initializer that generates tensors with a uniform distribution.</li>
<li>
<strong>TruncatedNormal</strong>: Initializer that generates a truncated normal distribution.</li>
<li>
<strong>VarianceScaling</strong>: Initializer capable of adapting its scale to the shape of weights.</li>
<li>
<strong>Orthogonal</strong>: Initializer that generates a random orthogonal matrix.</li>
<li>
<strong>Identity</strong>: Initializer that generates the identity matrix.</li>
<li>
<strong>lecun_uniform</strong>: LeCun uniform initializer.</li>
<li>
<strong>glorot_normal</strong>: Glorot normal initializer, also called Xavier normal initializer.</li>
<li>
<strong>glorot_uniform</strong>: Glorot uniform initializer, also called Xavier uniform initializer.</li>
<li>
<strong>he_normal</strong>: He normal initializer.</li>
<li>
<strong>lecun_normal</strong>: LeCun normal initializer.</li>
<li>
<strong>he_uniform</strong>: He uniform variance scaling initializer.</li>
</ul>
<p>See the <a href="https://keras.io/initializers/">documentation</a> for more details.</p>
<p>Out of interest, the default initializers chosen by Keras developers for different layer types are as follows:</p>
<ul>
<li>
<strong>Dense</strong> (e.g. MLP): <em>glorot_uniform</em>
</li>
<li>
<strong>LSTM</strong>: <em>glorot_uniform</em>
</li>
<li>
<strong>CNN</strong>: <em>glorot_uniform</em>
</li>
</ul>
<p>You can learn more about &ldquo;<em>glorot_uniform</em>&ldquo;, also called &ldquo;<em>Xavier normal</em>&ldquo;, named for the developer of the method Xavier Glorot, in the paper:</p>
<ul>
<li>
<a href="http://proceedings.mlr.press/v9/glorot10a.html">Understanding the difficulty of training deep feedforward neural networks</a>, 2010.</li>
</ul>
<p>There is no single best way to initialize the weights of a neural network.</p>
<blockquote>
<p>Modern initialization strategies are simple and heuristic. Designing improved initialization strategies is a difficult task because neural network optimization is not yet well understood. [&hellip;] Our understanding of how the initial point affects generalization is especially primitive, offering little to no guidance for how to select the initial point.</p>
</blockquote>
<p>&mdash; Page 301, <a href="https://amzn.to/2H5wjfg">Deep Learning</a>, 2016.</p>
<p>It is one more hyperparameter for you to explore and test and experiment with on your specific predictive modeling problem.</p>
<p>Do you have a favorite method for weight initialization?<br>
Let me know in the comments below.</p>
<h2>Further Reading</h2>
<p>This section provides more resources on the topic if you are looking to go deeper.</p>
<h3>Books</h3>
<ul>
<li>
<a href="https://amzn.to/2H5wjfg">Deep Learning</a>, 2016.</li>
</ul>
<h3>Articles</h3>
<ul>
<li><a href="https://en.wikipedia.org/wiki/Nondeterministic_algorithm">Nondeterministic algorithm on Wikipedia</a></li>
<li><a href="https://en.wikipedia.org/wiki/Randomized_algorithm">Randomized algorithm on Wikipedia</a></li>
<li><a href="https://en.wikipedia.org/wiki/Stochastic_optimization">Stochastic optimization on Wikipedia</a></li>
<li><a href="https://en.wikipedia.org/wiki/Stochastic_gradient_descent">Stochastic gradient descent on Wikipedia</a></li>
<li><a href="https://en.wikipedia.org/wiki/Fitness_landscape">Fitness landscape on Wikipedia</a></li>
<li><a>Neural Network FAQ</a></li>
<li><a href="https://keras.io/initializers/">Keras Weight Initialization</a></li>
<li>
<a href="http://proceedings.mlr.press/v9/glorot10a.html">Understanding the difficulty of training deep feedforward neural networks</a>, 2010.</li>
</ul>
<h3>Discussion</h3>
<ul>
<li><a href="https://stats.stackexchange.com/questions/47590/what-are-good-initial-weights-in-a-neural-network">What are good initial weights in a neural network?</a></li>
<li><a href="https://stackoverflow.com/questions/20027598/why-should-weights-of-neural-networks-be-initialized-to-random-numbers">Why should weights of Neural Networks be initialized to random numbers?</a></li>
<li><a href="https://www.quora.com/What-are-good-initial-weights-in-a-neural-network">What are good initial weights in a neural network?</a></li>
</ul>
<h2>Summary</h2>
<p>In this post, you discovered why neural network weights must be randomly initialized.</p>
<p>Specifically, you learned:</p>
<ul>
<li>About the need for nondeterministic and randomized algorithms for challenging problems.</li>
<li>The use of randomness during initialization and search in stochastic optimization algorithms.</li>
<li>That stochastic gradient descent is a stochastic optimization algorithm and requires the random initialization of network weights.</li>
</ul>
<p>Do you have any questions?<br>
Ask your questions in the comments below and I will do my best to answer.</p>
<p>The post <a rel="nofollow" href="https://machinelearningmastery.com/why-initialize-a-neural-network-with-random-weights/">Why Initialize a Neural Network with Random Weights?</a> appeared first on <a rel="nofollow" href="https://machinelearningmastery.com">Machine Learning Mastery</a>.</p>
<br>
<br>
<br>
<br>
AI|ML|DM|IR|PR<br>
<br>
via Machine Learning Mastery https://ift.tt/2if3vo8<br>
<br>
August 1, 2018 at 03:12AM
